[prometheus-redis-exporter] support multiple targets

LeoQuote opened the pull request:
<!--
Thank you for contributing to prometheus-community/helm-charts.
Before you submit this PR we'd like to make sure you are aware of our technical requirements and best practices:

* https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#technical-requirements
* https://helm.sh/docs/chart_best_practices/

For a quick overview across what we will look at reviewing your PR, please read our review guidelines:

// TODO: add a REVIEW_GUIDELINES.md in prometheus-community/helm-charts
* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the history.
This will make it easier to identify new changes.
The PR will be squashed anyways when it is merged.
Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them.
Once pushed, GitHub Actions will run across your changes and do some initial checks and linting.
These checks run very quickly.
Please check the results.
We would like these checks to pass before we even continue reviewing your changes.
-->
#### What this PR does / why we need it:

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #2198 

#### Special notes for your reviewer:

#### Checklist
<!-- [Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

monotek reviewed (changes requested):
You have multiple breaking changes in this pr:
* default labels change -> whats the upgrade path?
* customLabels were removed -> why?
* service port variable changed -> why?


You also have to sign your commits.

LeoQuote commented:
@monotek will do, thanks for the review

LeoQuote commented:
@monotek 

default labels was changed, but `managed-by: Helm` label is still present, helm upgrade can handle the upgrade, I added a note in readme to notify people that selectorLabels are changed.

I also added detailed comments in values.yaml to clearify where customLabels and labels will be used.

service port variable is changed to service port name , I think this could be more Intuitive, targetPort is actually container port, sometimes this could be misleading.

monotek reviewed (changes requested):
Upgrade from 4.8.0 fails. Please provide upgradepath.

```
W0629 10:27:05.881840  124326 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0629 10:27:05.882859  124326 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0629 10:27:05.887939  124326 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
Error: UPGRADE FAILED: cannot patch "prom-redis-prometheus-redis-exporter" with kind Deployment: Deployment.apps "prom-redis-prometheus-redis-exporter" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/instance":"prom-redis", "app.kubernetes.io/name":"prometheus-redis-exporter"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable

```

monotek commented on charts/prometheus-redis-exporter/README.md:
````diff
@@ -50,6 +50,42 @@ $ helm upgrade [RELEASE_NAME] [CHART] --install
+
+
+```console
+helm uninstall {{ .Release.Name }}
````
You could keep your helm history if you delete the deployment before upgrading:

```suggestion
kubectl delete deployment -l app=prometheus-redis-exporter
helm upgrade -i prometheus-redis-exporter prometheus-community/prometheus-redis-exporter
```

LeoQuote replied:
Thank you! I didn't know that could be done

monotek commented on charts/prometheus-redis-exporter/README.md:
```diff
@@ -50,6 +50,42 @@ $ helm upgrade [RELEASE_NAME] [CHART] --install
 
+### To 5.0.0
+
+From 5.0.0 redis exporter is using this above as selector labels in service and deployments
```
```suggestion
From 5.0.0 redis exporter is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therfore youl have to delete the deployment before your upgrade.
```

monotek commented on charts/prometheus-redis-exporter/README.md:
````diff
@@ -50,6 +50,42 @@ $ helm upgrade [RELEASE_NAME] [CHART] --install
+### To 5.0.0
+
+From 5.0.0 redis exporter is using this above as selector labels in service and deployments
+```
````
Line 56 - 62 can be deleted.

LeoQuote replied:
It is deleted now.

monotek commented on charts/prometheus-redis-exporter/README.md:
```diff
@@ -50,6 +50,35 @@ $ helm upgrade [RELEASE_NAME] [CHART] --install
 
+### To 5.0.0
+
+From 5.0.0 redis exporter is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therfore youl have to delete the deployment before your upgrade.
```
Sorry, some typos...

```suggestion
From 5.0.0 redis exporter is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therefore you have to delete the deployment before you upgrade.
```

LeoQuote commented:
markdown lint found serveral issues, but these seems not to be real issue , 

`helm repo` and `github` are very typical usage , especially in this repo

LeoQuote commented:
Thanks for your edit ! @monotek I think this chart could work now

monotek commented:
Let's wait for the reviews of @acondrat or @zanhsieh as they are codeowners.

Sameer1808 commented:
@LeoQuote the multiple targets thing still isn't working for me, here is the config for the exporter :

# If serviceMonitor.multipleTarget is enabled, this configuration is actually not used
#redisAddress: redis://<redis-server-1>:6379


serviceMonitor:
  # When set true then use a ServiceMonitor to configure scraping
  enabled: true
  multipleTarget: true
  telemetryPath: /scrape
  targets:
  # for every targets, url and name must be set,
  # an individual additionalRelabeling can be set for every target
  - url: redis://<redis-server-1>:6379
    name: redis-server-1
  - url: redis://<redis-server-2>:6379
    name: redis-server-2
  additionalMetricsRelabels: {}
  additionalRelabeling: []


we have tried with and without passing the telemetry path but still there are no metrics in prometheus, we have checked the connectivity..
its working if we pass single host in `redisAddress` .. let us know if we are missing anything here

LeoQuote commented:
@Sameer1808 hi, would your open a new issue and refer to this pull request?
more info is required to debug this problem , like the redis exporter image you’re using, full values file if convenient.

shubhamsetia05 commented:
@LeoQuote as per your suggestion, a new issue has been opened - https://github.com/prometheus-community/helm-charts/pull/2199, also attached the values file we are using.
